text/0000-serialize-deserialize.md

```diff
@@ -0,0 +1,174 @@
+
+This can make testing difficult, especially when running browser tests since there is not
+an easy way to mock the deserialization other than each plugin exposing their own deserialization
+token. 
+
+Providing a standard API for sharing json data between server and browser plugins
+will make it easier to test these types of plugins in isolation without needing to resort 
+to integration tests. Additionally, it can help solve problems where users may want custom 
+serialization / deserialization mechanisms, such as using tools like immutable.js.
+
+# Detailed design
+
+This proposal adds the following APIs to `fusion-core` 
+
+### `SharedStateToken`
+
+```js
+import {SharedStateToken} from 'fusion-core';
+```
+
+The `SharedStateToken` will have a default implementation registered onto each fusion app
+which provides a `{serialize}` and `{deserialize}` function which serialize json data on the
+server into a `<script type='application/json'>` dom element which is appended to `ctx.template.body`, and
+deserialize the data from the dom node in the browser. Example Usage:
+
+```js
+import {SharedStateToken, createPlugin} from 'fusion-core';
+
+export default createPlugin({
+  deps: {
+    SharedState: SharedStateToken,
+  },
+  middleware: ({SharedState}) => {
+    const {serialize, deserialize} = SharedState('some-id');
+    return (ctx, next) => {
+      if (__BROWSER__) {
+        // load and parse serialized data from dom node
+        const data = deserialize();
+      } else {
+        // serialize json data into body
+        serialize(ctx, {
+          some: 'json',
+        });
+      }
+      return next();
+    };
+  },
+});
+```
+
+### `MockSharedState`
+
+```js
+import {MockSharedState} from 'fusion-core';
+```
+
+This API will provide an easy way to provide mock data for all deserialize
+calls in an application. 
+
+```js
+import {MockSharedState, SharedStateToken} from 'fusion-core';
+// Deserialize calls will now return the mock values for these ids
+app.register(SharedStateToken, MockSharedState({
+  'some-id': {
+    some: 'json'
+  },
+  'other-data': {}
+}));
+```
+
+### Custom Serialization
+
+This structure can also be useful when trying to implement custom serialization / deserialization for certain plugins. 
+For example, lets say you want to use the `fusion-plugin-react-redux` plugin but want to have custom serialization / deserialization
+to support immutable-js in your store. You can create your own custom serializer/deserializer, and even set up your app to only
+use this custom SharedState implementation for the react redux plugin using token aliasing.
+
+```js
+import {createToken, html, SharedStateToken} from 'fusion-core';
+import ReduxPlugin, {ReduxToken} from 'fusion-plugin-react-redux';
+import transit from 'transit-immutable-js';
+
+const ImmutableSharedState = (id) => {
+  return {
+    serialize: (ctx, data) => {
+      ctx.template.body.push(
```

Comment:
Why are you writing to and reading from HTML directly? I'd rather have the framework abstract this away for me.

I understand you want an escape hatch, but I think the default case should be that the framework sets the return value of this function in the DOM, and then on the browser is responsible for trying to find that element and passing its results as an argument to deserialize.

Reply:
By default the framework handles writing to html / dom. It is only in the case where you want custom serialization / deserialization when you need to do this manually. 

The reason I think we need the escape hatch to include html/dom APIs is it could be dependent on your serialization mechanism. For example, you might want to serialize it into an inline js script rather than a json script tag.